Introduce Nuage event catcher

gberginc opened the pull request:
This commit introduces the full blown event catcher for Nuage manager based on ActiveMQ message bus. Further to default authentication, AMQP is now provided via Qpid Proton gem, along with the corresponding message handler responsible for the delivery of messages into CFME.

gberginc commented:
@miq-bot add_label wip

The reason for WIP is that it requires Qpid Proton gem that is currently not properly available (it also requires host libs). Hopefully, the required version will be ready in the second half of September!

In order to test this, I am using the following gem in my main Gemfile
```
gem "qpid_proton", "~>0.17.0.1", :git => "https://github.com/gberginc/qpid_proton_gem"
```

Dry code review would be much appreciated to simplify merging afterwards.

@miha-plesko Would you be willing to make the first pass?

gberginc commented on app/models/manageiq/providers/nuage/network_manager/event_catcher/runner.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  def topic_names
+    "topic://topic/CNAMessages"
```
Yes, I am discussing this with Nuage currently. I think using settings will be ideal, in particular because there are two at least two topics (events and alerts) that are of interest.

gberginc commented on app/models/manageiq/providers/nuage/network_manager/event_catcher/messaging_handler.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def on_connection_error(event)
+    raise MiqException::MiqInvalidCredentialsError.new "Connection failed due to bad username or password"
```
From what I have seen, the `event` does not contain additional information. I've seen CC comments and will address them

gberginc commented on app/models/manageiq/providers/nuage/manager_mixin.rb:
```diff
@@ -41,12 +41,46 @@ def translate_exception(err)
+  def auth_url(protocol, server, port, version)
+    scheme = protocol == "ssl-with-validation" ? "https" : "http"
+    "#{scheme}://#{server}:#{port}/nuage/api/#{version}"
+  end
```
No, I incorrectly moved it while introducing new functionalities

gberginc commented on app/models/manageiq/providers/nuage/network_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,46 @@
+  def self.test_amqp_connection(options = {})
+    # Ensure we just test the connection. AMQP channel will be established and
+    # started, however it will be immediately stopped.
+    options[:test_connection] = true
```
Yes, unfortunately it works like this. In case the URL is incorrect, this will be noticed immediately when opening the connection, but in case just the `user:pass` is incorrect, we need to wait for the connection to be opened - otherwise the corresponding error event will be thrown.

gberginc commented on app/models/manageiq/providers/nuage/network_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,46 @@
+
+  def initialize(options = {})
+    @options           = options
+    @collecting_events = false
```
Nah, that is your fault as I followed your [example](https://github.com/ManageIQ/manageiq-providers-vmware/blob/ba1583e269620784bdbe5e367e80fed61fc95f5d/app/models/manageiq/providers/vmware/cloud_manager/event_catcher/stream.rb#L36) 😜

gberginc commented:
Thanks @miha-plesko! All comments addressed.

CC now happy, but now rubocop says I have set everything on fire 😄.

@juliancheal the two lines identified as `unexpected token error` were changed from

```
@message_handler_block.call(JSON.parse(event.message.body)) if @message_handler_block
@event_monitor_handle.try!(:stop)
```

because codeclimate suggested the use of `safe navigation (&.)`. I am fine either way, depends mainly on what the target Ruby version is currently (safe navigation has been around since 2.3, but seems like rubocop is not configured for that).

gberginc commented:
@miq-bot add_label enhancement,events

miha-plesko commented on app/models/manageiq/providers/nuage/network_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,47 @@
+  end
+
+  def stop
+    @handler.stop
```
I understand `@handler.stop` sets `@stop_listening = true` which is then read in `on_message` function of the MessagingHandler. I'm wondering if this is sufficient - if no new message arrives after stopping, then the @handler will keep running.

I'm not familiar with qpid proton gem, but would it work if we use something like:

```ruby
@handler.interrupt
```

to stop the event catcher? See [this](https://github.com/gberginc/qpid_proton_gem/blob/6188f0732382fc7983f5c0aa8f6da379050d71fd/lib/messenger/messenger.rb#L399-L413) documentation. Please note, it's just an idea of mine, not sure if it's doable.

gberginc replied:
Problem with this is that it's again part of the `messenger`, not the `reactor`. The reactor was chosen for this use case as it seems more lightweight as we are only listening to the events.

I understand your concern and will investigate the reactor to see if there is similar way to stop  the handler

gberginc replied:
@miha-plesko Check [this commit](https://github.com/ManageIQ/manageiq-providers-nuage/pull/9/commits/f5707bc13f6a3142390abbe0236c3d6f1b097cc4) please. This now automatically closes the connection when stop is requested. Thanks for pointing out the problem.

miha-plesko replied:
I like it better now, thanks for this change.

gberginc commented:
@miha-plesko I've added a [patch](https://github.com/ManageIQ/manageiq-providers-nuage/pull/9/commits/9f1124be95b6ba406515285d1fd8fc9806b7c9a6) that uses event catcher settings to specify the topics of interest. This allows us to enable standard events as well as alarms from the config file, with no change in the code. With the previous commit that handles connection closing immediately, it makes it much more practical.

gberginc commented:
@juliancheal Do you perhaps have some time to review the code, despite the fact that we don't have the correct gem in place? Please note that the worker is not enabled by default (it's not yet added [here](https://github.com/ManageIQ/manageiq/blob/master/lib/workers/miq_worker_types.rb#L32)) so this PR does not actually break anything.

miha-plesko commented on app/models/manageiq/providers/nuage/manager_mixin.rb:
```diff
@@ -41,12 +41,41 @@ def translate_exception(err)
+      {
+        :ems                       => self,
+        :url                       => url,
+        :sasl_allow_insecure_mechs => true, # Only plain (insecure) mechanism currently supported
```
Now that we've introduced settings.yaml, we could probably also move `:sasl_allow_insecure_mechs` decision in there.

gberginc replied:
Insecure will come from the AMQP settings from the UI (you probably remember the "new provider" where you have an option for secure/not secure).

The reason to have this hardcoded is simply because Nuage supposedly only uses amqp and not amqps - thus, for now even the UI will not allow ssl.

miha-plesko replied:
Ah, yes, you are right. Thanks for the explanation!

sergio-ocon commented:
How is the status of the library needed?
Would this work in other places?

gberginc commented:
I haven't seen the update from qpid_proton - will post a question.

From what I was able to test, this PR is now in a shape that it should allow merging without any disruption in the existing provider. Eventing will not be enabled unless a change in core is also made.

juliancheal commented:
> Thanks @miha-plesko! All comments addressed.
> 
> CC now happy, but now rubocop says I have set everything on fire 😄.
> 
> @juliancheal the two lines identified as unexpected token error were changed from
> 
> @message_handler_block.call(JSON.parse(event.message.body)) if @message_handler_block
> @event_monitor_handle.try!(:stop)
> because codeclimate suggested the use of safe navigation (&.). I am fine either way, depends mainly on what the target Ruby version is currently (safe navigation has been around since 2.3, but seems like rubocop is not configured for that).

I think we are still supporting 2.2.2, so wouldn't be able to use `&.`. I remember this coming up with a PR from Lenovo. @Fryguy would would you say.

juliancheal commented:
I'm good with this PR, except the `&.`, but want to see if @Fryguy agrees.

What's stopping the correct gem being in place to be used by this PR?

gberginc commented:
@juliancheal [this](https://github.com/apache/qpid-proton/pull/116#issuecomment-331201529) is the latest response from Qpid team. Problem is that current version ([0.17](https://rubygems.org/gems/qpid_proton/versions/0.17.0)) has a bug preventing Nuage provider to connect to ActiveMQ. We have fixed this (the above PR explains in some more detail) and are currently using a custom Gem with the patch applied.

As written above, the event catcher currently does not affect anything because it is not enabled by default. I am thus fine removing the WIP in case you are also fine accepting the code in such state. Once new Gem is released, we will be able to enable it in the list of workers (core repo).

juliancheal commented:
@gberginc thanks for the clarification 👍

sergio-ocon commented:
What happened to the underlying libraries needed for this to work?
Do we need changes in the build to make this work?

gberginc commented:
@juliancheal Since I have reverted the use of "safe navigation", this again causes codeclimate to fail, however I believe is is still much better than those explosive rubocop warnings about older Ruby versions. Also, it's in line with the code of other similar providers.

I am now also removing the wip label as it should be safe for the merge.

@miq-bot remove_label wip

blomquisg commented on app/models/manageiq/providers/nuage/network_manager/event_parser.rb:
```diff
@@ -0,0 +1,13 @@
+module ManageIQ::Providers::Nuage::NetworkManager::EventParser
+  def self.event_to_hash(event, ems_id)
+    {
+      :source     => "Nuage",
```
For consistency sake, please change to "NUAGE".

gberginc commented:
Thank you very much @blomquisg for taking the time to review the code. I've changed the source as you suggested and am now waiting for @durandom.

sergio-ocon commented:
Code climate is complaining for not using &., but we asked them to do so before to be able to use ruby 2.2

@durandom ?

durandom commented on app/models/manageiq/providers/nuage/network_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,45 @@
+require 'qpid_proton'
```
the `qpid_proton` gem is nowhere included. AFAIK we dont use it yet in the core.

@kbrock when / will we include it in the cores Gemfile?

If not you'll have to add it to the providers gemspec

durandom replied:
@gberginc  if you add a spec, that will trigger this code path, I guess it will throw an exception

Fryguy replied:
> If not you'll have to add it to the providers gemspec

It should be included in the provider's gemspec regardless, because the provider should always declare it's dependencies.  This way if we, say, have to bump the version in the main repo, we can ensure we pick a version that satisfies the dependencies of this plugin

Fryguy replied:
Same...this still cannot be at the top-level context

Fryguy replied:
I don't think you even need this as long as the MessagingHandler does the require in it's initialize method.

durandom commented on app/models/manageiq/providers/nuage/network_manager/event_parser.rb:
```diff
@@ -0,0 +1,13 @@
+      :source     => "NUAGE",
+      :event_type => "#{event['entityType']}_#{event['type'].downcase}",
+      :timestamp  => DateTime.strptime((event["eventReceivedTime"] / 1000).to_s, '%s').to_s,
+      :message    => event.to_hash,
```
I think `message` should rather be a single string and `full_data` contains the complete hash.

durandom commented on app/models/manageiq/providers/nuage/network_manager/event_parser.rb:
```diff
@@ -0,0 +1,13 @@
+      :timestamp  => DateTime.strptime((event["eventReceivedTime"] / 1000).to_s, '%s').to_s,
+      :message    => event.to_hash,
+      :vm_ems_ref => nil,
+      :full_data  => event,
```
I guess this will be a hash?
But maybe, just to be sure, when something in qpid changes and they refactor to use a object instead, you'll want to make this `event.to_hash`

gberginc commented:
@durandom Thanks!

You were not in the emails we are exchanging with Satoe about Qpid gem. Two things are required: first, few [system packages](https://github.com/ManageIQ/manageiq-appliance-build/pull/228/files), and then install the gem from a forked [Qpid repo](https://github.com/gberginc/qpid_proton_gem). Problem with the latter is that Qpid has not released an update that is mandatory for this to work with Nuage.

I have prepared a [simple demo repo](https://github.com/gberginc/nuage_amqp_client) to demonstrate how it works.

Can you please advise what I should do? Should I propose to introduce the gem into the main Gemfile?

gberginc commented:
@durandom The [latest patch](https://github.com/ManageIQ/manageiq-providers-nuage/pull/9/commits/5f07e22955878ca078c52eda5305ab5324ec4c78) addresses the comments in the code.

Now I need to resolve the Qpid thing...

If other parts are fine, is it possible to get this in and then submit specs (which we are already working on) separately?

durandom commented:
> Can you please advise what I should do? Should I propose to introduce the gem into the main Gemfile?

@kbrock @Fryguy how should we handle the qpid dependency?
I'd say include it in this providers gemspec

gberginc commented:
Problem with provider's gemspec is that bundler does not allow github repositories. Thus, currently, the only way it is to get it inside main Gemfile. Once Qpid gets new build in RubyGems, provider's gemspec is the target.

juliancheal commented:
> Code climate is complaining for not using &., but we asked them to do so before to be able to use ruby 2.2

We asked Code Climate you mean? Yeah it's just for now while we support 2.2

juliancheal commented:
What's the status of the Qpid dependency?

gberginc commented:
The appliance-build has green light for [this PR](https://github.com/ManageIQ/manageiq-appliance-build/pull/228/files) (confirmed via email so I expect it to be merged soon).

The gem is still officially at 0.17, however the authors are working on releasing 0.18.

Fryguy commented:
Merged https://github.com/ManageIQ/manageiq-appliance-build/pull/228

Fryguy commented:
@gberginc What are you changing in the gem that's requiring you to use a git repo?

gberginc commented:
@Fryguy I can add the `qpid_proton` gem into the Nuage gemspec and it will pass most of the tests we have already in place offline. The problem with the current published version (0.17) is that it does not work with the authentication method Nuage is using (i.e. PLAIN). This means that using the current version, it will not connect to Nuage AMQP.

Therefore, I created a separated repo to host the 0.17 + the patch that was implemented based on the report I have provided to Qpid team.

Fryguy commented:
Ah ok...I misunderstood...from the way the appliance-build PR went, I thought that PLAIN was just a matter of adding the cyrus_sasl_plain package.

gberginc commented:
I've added a second PR (#10) that includes `qpid_proton` in gemspec (0.17, but for the tests this does not matter). However, it does not get installed due to a [missing system library](https://travis-ci.org/ManageIQ/manageiq-providers-nuage/jobs/283036285#L832).

Will the change to `manageiq-appliance-build` resolve this at some point?

durandom commented:
> Will the change to manageiq-appliance-build resolve this at some point?

afaik, the appliance stuff is not included for travis builds.

So, from my understanding, adding qpid and the .travis.yml changes to this PR here would make it mergeable.

sergio-ocon commented:
@gberginc Have those changes been done?

sergio-ocon commented:
The gem was supposed to be released already.
I've replied . to your request and Philipp's about this.

gberginc commented:
Thank you @sergio-ocon! Please, also see my last question in #11. I really would appreciate feedback to resolve this problem.

chargio commented:
@durandom @Fryguy Can you please have another look at this? With the PR that have been merged for the qpid gem, we should be ready to merge this.
If there is anything outstanding, please let us know so we can fix it this week.

Thanks

gberginc commented:
@durandom Thank you very much for you review!

The way I understand gemspec, it's not possible to add a gem from GitHub repository into gemspec. Once the gem will be released officially, I will add that as a dependency in Nuage provider.

I will also add the guides, yes. I just need to make sure to try every platform (Fedora, Ubuntu and mac OS).

gberginc commented:
@miq-bot assign @Fryguy 

This is now green based on the latest changes.

chargio commented:
@gberginc Is there any pending comment to address or are all them addressed?

juliancheal commented:
Seems like everything is done, as quid gem is installed now via other prs. @Fryguy are all your issues resolved? I'll merge this if so.

gberginc commented:
@chargio as far as I am concerned this PR is ready. Others, simpler, will follow.

Fryguy commented on app/models/manageiq/providers/nuage/network_manager/event_catcher/messaging_handler.rb:
```diff
@@ -0,0 +1,41 @@
+require 'qpid_proton'
```
This cannot be in the top-level context.  As the gem is not required to be installed, this class will blow up on load (even if the worker is not started).  You can avoid this by just putting the require in the initialize call.

Fryguy commented on app/models/manageiq/providers/nuage/network_manager/event_catcher/stream.rb:
```diff
@@ -0,0 +1,45 @@
+    # started, however it will be immediately stopped.
+    options[:test_connection] = true
+    begin
+      stream = ManageIQ::Providers::Nuage::NetworkManager::EventCatcher::Stream.new(options)
```
Minor, but you don't need to retype out the class name here...

```ruby
stream = new
stream.connection.run
```

gberginc commented:
@Fryguy just responding here as well in case you miss the gitter comment. Thanks for the review.

chargio commented:
Thanks @gberginc. @Fryguy it is supposed to be approved by every body now

Fryguy commented:
Need this one addressed: https://github.com/ManageIQ/manageiq-providers-nuage/pull/9#discussion_r146873094

After that, this is good to go!

juliancheal commented:
> Need this one addressed: #9 (comment)
> 
> After that, this is good to go!

Excellent, as soon as that change is in I'll press the big green button (which is actually grey) /cc @gberginc

juliancheal commented:
Cool beans. Just waiting on Travis.

gberginc commented:
/me starts the 🕺 😄.

chargio commented:
🥇
